File: apps/desktop/electron/db/redis.ts
import { createClient, RedisClientType } from 'redis'
import { DatabaseConnection, ConnectionConfig, StorageAnalysis, QueryResult, SchemaInfo } from './index'
import { writeFileSync } from 'fs'

export class RedisConnection extends DatabaseConnection {
  private client?: RedisClientType

  async connect(): Promise<void> {
    const connectionString = this.config.connectionString || 
      `redis://${this.config.username ? `${this.config.username}:${this.config.password}@` : ''}${this.config.host}:${this.config.port || 6379}`

    this.client = createClient({ url: connectionString })
    await this.client.connect()
    this.connected = true
  }

  async disconnect(): Promise<void> {
    if (this.client) {
      await this.client.quit()
      this.client = undefined
      this.connected = false
    }
  }

  async testConnection(): Promise<void> {
    if (!this.client) {
      await this.connect()
    }
    
    await this.client!.ping()
  }

  async analyzeStorage(): Promise<StorageAnalysis> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    // Get Redis info
    const info = await client.info('memory')
    const memoryInfo = this.parseRedisInfo(info)
    const totalSize = memoryInfo.used_memory || 0

    // Get database info
    const dbInfo = await client.info('keyspace')
    const keyspaceInfo = this.parseRedisInfo(dbInfo)

    // Get all keys (this can be expensive for large databases)
    const keys = await client.keys('*')
    const keyCount = keys.length

    // Analyze key types and sizes
    const keyAnalysis = new Map<string, { count: number; totalSize: number }>()
    
    for (const key of keys.slice(0, 1000)) { // Limit to first 1000 keys for performance
      try {
        const type = await client.type(key)
        const memoryUsage = await client.memoryUsage(key)
        
        if (!keyAnalysis.has(type)) {
          keyAnalysis.set(type, { count: 0, totalSize: 0 })
        }
        
        const analysis = keyAnalysis.get(type)!
        analysis.count++
        analysis.totalSize += memoryUsage
      } catch (error) {
        // Skip keys that can't be analyzed
      }
    }

    // Convert to tables format
    const tables = Array.from(keyAnalysis.entries()).map(([type, analysis]) => ({
      name: `keys:${type}`,
      size: analysis.totalSize,
      rowCount: analysis.count,
      indexSize: 0,
      bloat: 0
    }))

    // Sort by size descending
    tables.sort((a, b) => b.size - a.size)

    const largestTable = tables.length > 0 ? tables[0] : { name: 'N/A', size: 0, rowCount: 0 }

    return {
      totalSize,
      tableCount: tables.length,
      indexCount: 0, // Redis doesn't have traditional indexes
      largestTable,
      tables,
      indexes: [], // Redis doesn't have traditional indexes
      lastAnalyzed: new Date().toISOString()
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!
    const startTime = Date.now()

    try {
      // Parse Redis command (simplified)
      const parts = query.trim().split(' ')
      const command = parts[0].toUpperCase()
      const args = parts.slice(1)

      let result: any

      switch (command) {
        case 'GET':
          result = await client.get(args[0])
          break
        case 'SET':
          result = await client.set(args[0], args[1])
          break
        case 'KEYS':
          result = await client.keys(args[0] || '*')
          break
        case 'INFO':
          result = await client.info(args[0] || 'server')
          break
        case 'PING':
          result = await client.ping()
          break
        case 'DBSIZE':
          result = await client.dbSize()
          break
        case 'FLUSHDB':
          result = await client.flushDb()
          break
        case 'FLUSHALL':
          result = await client.flushAll()
          break
        default:
          // Try to execute as a generic command
          result = await client.sendCommand([command, ...args])
      }

      const executionTime = Date.now() - startTime

      // Format result for display
      const rows = Array.isArray(result) ? result.map(item => ({ value: item })) : [{ value: result }]
      const columns = ['value']

      return {
        columns,
        rows,
        rowCount: rows.length,
        executionTime,
        explainPlan: undefined
      }
    } catch (error) {
      throw new Error(`Redis command failed: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async getSchema(): Promise<SchemaInfo> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    // Get all keys and their types
    const keys = await client.keys('*')
    const keyTypes = new Map<string, number>()

    for (const key of keys.slice(0, 1000)) { // Limit for performance
      try {
        const type = await client.type(key)
        keyTypes.set(type, (keyTypes.get(type) || 0) + 1)
      } catch (error) {
        // Skip keys that can't be analyzed
      }
    }

    // Convert to schema format
    const tables = Array.from(keyTypes.entries()).map(([type, count]) => ({
      name: `keys:${type}`,
      columns: [
        { name: 'key', type: 'string', nullable: false, defaultValue: undefined },
        { name: 'value', type: type, nullable: true, defaultValue: undefined }
      ],
      indexes: []
    }))

    return {
      schemas: [{
        name: 'default',
        tables
      }]
    }
  }

  async createBackup(backupPath: string): Promise<{ path: string; size: number }> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    try {
      // Get all keys and their values
      const keys = await client.keys('*')
      const backupData: any = {
        database: this.config.database || 'default',
        keys: {}
      }

      for (const key of keys) {
        try {
          const type = await client.type(key)
          let value: any

          switch (type) {
            case 'string':
              value = await client.get(key)
              break
            case 'list':
              value = await client.lRange(key, 0, -1)
              break
            case 'set':
              value = await client.sMembers(key)
              break
            case 'zset':
              value = await client.zRangeWithScores(key, 0, -1)
              break
            case 'hash':
              value = await client.hGetAll(key)
              break
            default:
              value = null
          }

          backupData.keys[key] = { type, value }
        } catch (error) {
          // Skip keys that can't be backed up
        }
      }

      // Write backup to file
      const backupJson = JSON.stringify(backupData, null, 2)
      writeFileSync(backupPath, backupJson)

      const stats = require('fs').statSync(backupPath)
      return { path: backupPath, size: stats.size }
    } catch (error) {
      throw new Error(`Failed to create Redis backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async restoreBackup(backupPath: string): Promise<void> {
    if (!this.client) {
      await this.connect()
    }

    const client = this.client!

    try {
      // Read backup file
      const backupData = JSON.parse(require('fs').readFileSync(backupPath, 'utf8'))

      // Clear existing database
      await client.flushDb()

      // Restore keys
      for (const [key, keyData] of Object.entries(backupData.keys)) {
        const { type, value } = keyData as any

        switch (type) {
          case 'string':
            await client.set(key, value)
            break
          case 'list':
            if (Array.isArray(value)) {
              await client.lPush(key, ...value)
            }
            break
          case 'set':
            if (Array.isArray(value)) {
              await client.sAdd(key, ...value)
            }
            break
          case 'zset':
            if (Array.isArray(value)) {
              await client.zAdd(key, value)
            }
            break
          case 'hash':
            if (typeof value === 'object' && value !== null) {
              await client.hSet(key, value)
            }
            break
        }
      }
    } catch (error) {
      throw new Error(`Failed to restore Redis backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  private parseRedisInfo(info: string): Record<string, any> {
    const result: Record<string, any> = {}
    const lines = info.split('\r\n')
    
    for (const line of lines) {
      if (line.includes(':')) {
        const [key, value] = line.split(':', 2)
        result[key] = isNaN(Number(value)) ? value : Number(value)
      }
    }
    
    return result
  }
}

File: db_storage_manager/db/redis.py
"""
Redis database connection
"""

import asyncio
import subprocess
from pathlib import Path
from typing import Any, Dict, List
import redis

from .base import (
    DatabaseConnection,
    ConnectionConfig,
    StorageAnalysis,
    QueryResult,
    SchemaInfo,
)


class RedisConnection(DatabaseConnection):
    """Redis database connection"""

    def __init__(self, config: ConnectionConfig):
        super().__init__(config)
        self.client = None

    async def connect(self) -> None:
        """Connect to Redis database"""
        if self.config.connection_string:
            self.client = redis.from_url(self.config.connection_string)
        else:
            self.client = redis.Redis(
                host=self.config.host or "localhost",
                port=self.config.port or 6379,
                db=int(self.config.database) if self.config.database else 0,
                password=self.config.password,
                decode_responses=False,
            )

        # Test connection
        self.client.ping()
        self.connected = True

    async def disconnect(self) -> None:
        """Disconnect from Redis database"""
        if self.client:
            self.client.close()
            self.client = None
        self.connected = False

    async def analyze_storage(self) -> StorageAnalysis:
        """Analyze Redis storage"""
        if not self.client:
            await self.connect()

        info = self.client.info("memory")
        total_size = info.get("used_memory", 0)

        # Get all keys
        keys = self.client.keys("*")
        key_types = {}
        key_sizes = {}

        for key in keys:
            key_type = (
                self.client.type(key).decode()
                if isinstance(self.client.type(key), bytes)
                else self.client.type(key)
            )
            key_types[key_type] = key_types.get(key_type, 0) + 1

            try:
                memory_usage = self.client.memory_usage(key)
                key_sizes[key] = memory_usage
            except Exception:
                key_sizes[key] = 0

        # Create tables from key types
        tables = []
        for key_type, count in key_types.items():
            tables.append(
                {
                    "name": f"keys:{key_type}",
                    "size": 0,
                    "rowCount": count,
                    "indexSize": 0,
                    "bloat": 0.0,
                }
            )

        # Find largest key
        largest_key = {"name": "N/A", "size": 0}
        if key_sizes:
            largest_key_name = max(key_sizes, key=key_sizes.get)
            largest_key = {
                "name": (
                    largest_key_name.decode()
                    if isinstance(largest_key_name, bytes)
                    else largest_key_name
                ),
                "size": key_sizes[largest_key_name],
            }

        return {
            "totalSize": total_size,
            "tableCount": len(tables),
            "indexCount": 0,
            "largestTable": largest_key,
            "tables": tables,
            "indexes": [],
            "lastAnalyzed": __import__("datetime").datetime.now().isoformat(),
        }

    async def execute_query(self, query: str, safe_mode: bool = True) -> QueryResult:
        """Execute a Redis command"""
        if not self.client:
            await self.connect()

        import time
        import json

        start_time = time.time()

        try:
            # Try to parse as JSON array for multi commands
            try:
                commands = json.loads(query)
                if isinstance(commands, list) and all(
                    isinstance(c, list) for c in commands
                ):
                    pipe = self.client.pipeline()
                    for cmd in commands:
                        pipe.execute_command(*cmd)
                    result = pipe.execute()
                else:
                    result = [self.client.execute_command(*commands)]
            except (json.JSONDecodeError, TypeError):
                # Parse as space-separated command
                parts = query.split()
                command = parts[0].upper()
                args = parts[1:] if len(parts) > 1 else []

                # Decode bytes arguments
                args = [arg.decode() if isinstance(arg, bytes) else arg for arg in args]

                result = self.client.execute_command(command, *args)

            # Normalize result
            if not isinstance(result, list):
                result = [result]

            # Decode bytes in result
            def decode_value(val):
                if isinstance(val, bytes):
                    try:
                        return val.decode()
                    except:
                        return val
                elif isinstance(val, list):
                    return [decode_value(v) for v in val]
                elif isinstance(val, dict):
                    return {k: decode_value(v) for k, v in val.items()}
                return val

            result = [decode_value(r) for r in result]

            execution_time = int((time.time() - start_time) * 1000)

            return {
                "columns": ["result"],
                "rows": [{"result": r} for r in result],
                "rowCount": len(result),
                "executionTime": execution_time,
                "explainPlan": None,
            }
        except Exception as e:
            raise RuntimeError(f"Query execution failed: {str(e)}")

    async def get_schema(self) -> SchemaInfo:
        """Get Redis schema"""
        if not self.client:
            await self.connect()

        keys = self.client.keys("*")
        key_types = {}

        for key in keys:
            key_type = (
                self.client.type(key).decode()
                if isinstance(self.client.type(key), bytes)
                else self.client.type(key)
            )
            key_types[key_type] = key_types.get(key_type, 0) + 1

        # Create tables from key types
        tables = []
        for key_type, count in key_types.items():
            tables.append(
                {
                    "name": f"keys:{key_type}",
                    "columns": [
                        {
                            "name": "key",
                            "type": "string",
                            "nullable": False,
                            "defaultValue": None,
                        },
                        {
                            "name": "value",
                            "type": key_type,
                            "nullable": True,
                            "defaultValue": None,
                        },
                    ],
                    "indexes": [],
                }
            )

        return {
            "schemas": [
                {
                    "name": self.config.database or "default",
                    "tables": tables,
                }
            ]
        }

    async def create_backup(self, backup_path: str) -> Dict[str, Any]:
        """Create Redis backup"""
        if not self.client:
            await self.connect()

        # Save Redis RDB file
        self.client.save()

        # Get Redis data directory and dbfilename
        config = self.client.config_get("dir")
        redis_dir = config.get("dir", "/var/lib/redis")
        dbfilename_config = self.client.config_get("dbfilename")
        dbfilename = dbfilename_config.get("dbfilename", "dump.rdb")

        rdb_path = Path(redis_dir) / dbfilename

        if not rdb_path.exists():
            raise FileNotFoundError(f"Redis RDB file not found: {rdb_path}")

        backup_file = Path(backup_path)
        backup_file.parent.mkdir(parents=True, exist_ok=True)

        # Copy RDB file
        import shutil

        shutil.copy2(rdb_path, backup_file)
        size = backup_file.stat().st_size

        return {"path": str(backup_file), "size": size}

    async def restore_backup(self, backup_path: str) -> None:
        """Restore Redis backup"""
        backup_file = Path(backup_path)
        if not backup_file.exists():
            raise FileNotFoundError(f"Backup file not found: {backup_path}")

        if not self.client:
            await self.connect()

        # Get Redis data directory and dbfilename
        config = self.client.config_get("dir")
        redis_dir = config.get("dir", "/var/lib/redis")
        dbfilename_config = self.client.config_get("dbfilename")
        dbfilename = dbfilename_config.get("dbfilename", "dump.rdb")

        rdb_path = Path(redis_dir) / dbfilename

        # Disconnect
        await self.disconnect()

        # Copy backup file to Redis data directory
        import shutil

        shutil.copy2(backup_file, rdb_path)

        # Note: Redis needs to be restarted to load the new RDB file
        # This is typically done by the system administrator or Docker container restart
